tfs/frame.py

```diff
@@ -284,35 +301,69 @@ def _element_is_list(element):
+    # The following are deal-breakers for the TFS format and would not,
+    # for instance, be accepted by MAD-X or MAD-NG
     if any(not isinstance(c, str) for c in data_frame.columns):
         LOGGER.debug(f"Some column-names are not of string-type, dataframe {info_str} is invalid.")
-        errmsg = "TFS-Columns need to be strings."
-        raise TfsFormatError(errmsg)
+        raise NonStringColumnNameError
 
     if any(" " in c for c in data_frame.columns):
         LOGGER.debug(f"Space(s) found in TFS columns, dataframe {info_str} is invalid")
-        errmsg = "TFS-Columns can not contain spaces."
-        raise TfsFormatError(errmsg)
+        raise SpaceinColumnNameError
+
+    # ----- Additional checks for MAD-X compatibility mode ----- #
+    if compatibility.lower() in ("madx", "mad-x"):
+        # Check that no boolean values are in the headers
+        if any(isinstance(header, bool) for header in data_frame.headers.values()):
+            LOGGER.debug(
+                f"Boolean values found in headers of dataframe {info_str}, which is incompatible with MAD-X."
+                "Change their types in order to keep compatibility with MAD-X."
+            )
+            errmsg = "TFS-Headers can not contain boolean values in MAD-X compatibility mode."
+            raise MADXCompatibilityError(errmsg)
+
+        # Check that no complex values are in the headers
+        if any(isinstance(header, complex) for header in data_frame.headers.values()):
+            LOGGER.debug(
+                f"Complex values found in headers of dataframe {info_str}, which is incompatible with MAD-X."
+                "Change their types in order to keep compatibility with MAD-X."
+            )
+            errmsg = "TFS-Headers can not contain complex values in MAD-X compatibility mode."
+            raise MADXCompatibilityError(errmsg)
+
+        # Check that the dataframe contains no boolean dtype columns
+        if any(pdtypes.is_bool_dtype(type_) for type_ in data_frame.dtypes):
+            LOGGER.debug(
+                f"Boolean dtype column found in dataframe {info_str}, which is incompatible with MAD-X."
+                "Change the column dtypes to keep compatibility with MAD-X."
+            )
+            errmsg = "TFS-Dataframe can not contain boolean dtype columns in MAD-X compatibility mode."
+            raise MADXCompatibilityError(errmsg)
+
+        # Check that the dataframe contains no complex dtype columns
+        if any(pdtypes.is_complex_dtype(type_) for type_ in data_frame.dtypes):
+            LOGGER.debug(
+                f"Complex dtype column found in dataframe {info_str}, which is incompatible with MAD-X."
+                "Change the column dtypes or split it into real and imaginary values to keep compatibility with MAD-X."
+            )
+            errmsg = "TFS-Dataframe can not contain complex dtype columns in MAD-X compatibility mode."
+            raise MADXCompatibilityError(errmsg)
```

Comment:
**FEATURE**

All of these are new. They are what allows us to distinguish between MAD-NG and MAD-X modes.

Plainly, once the work of implementing boolean / complex in reader and writer are done, we just need to **not** accept them in `MAD-X` mode.